File: src/main/java/com/qtech/bubbles/registry/Registry.java
package com.qtech.bubbles.registry;

import com.qtech.bubbles.common.IRegistryEntry;
import com.qtech.bubbles.common.ResourceEntry;
import com.qtech.bubbles.common.init.ObjectInit;
import com.qtech.bubbles.common.maps.SequencedHashMap;
import com.qtech.bubbles.event._common.SubscribeEvent;
import com.qtech.bubbles.event.bus.Bus;
import com.qtech.bubbles.event.bus.EventBus;
import org.jetbrains.annotations.NotNull;

import javax.annotation.Nullable;
import java.lang.reflect.Field;
import java.util.*;


@SuppressWarnings("unchecked")
public class Registry<T extends IRegistryEntry> {
    private final SequencedHashMap<ResourceEntry, T> registry = new SequencedHashMap<>();
    private final Class<T> type;
    private static final HashMap<Class<?>, Registry<?>> metricsMap = new HashMap<>();
    private final ResourceEntry registryName;

    protected Registry(@NotNull Class<T> clazz, ResourceEntry registryName) throws IllegalStateException {
        if (metricsMap.containsKey(clazz)) {
            throw new IllegalStateException();
        }

        this.registryName = registryName;
        this.type = clazz;
        Bus.getQBubblesEventBus().register(this);

        Registry.metricsMap.put(type, this);
    }

    public ResourceEntry getRegistryName() {
        return registryName;
    }

    @Deprecated
    public static <T extends RegistryEntry> Registry<T> create(@NotNull Class<T> clazz, ResourceEntry registryName) {
        return new Registry<>(clazz, registryName);
    }

    public static <T extends RegistryEntry> Registry<T> getRegistry(Class<T> objType) {
        return (Registry<T>) metricsMap.get(objType);
    }

    /**
     * Returns the registered instance from the given {@link ResourceEntry}
     *
     * @param key the namespaced key.
     * @return an registered instance of the type {@link T}.
     * @throws ClassCastException if the type is invalid.
     */
    public T get(ResourceEntry key) {
        if (!registry.containsKey(key)) {
            throw new IllegalArgumentException("Cannot find object for: " + key + " | type: " + type.getSimpleName());
        }
        return registry.get(key);
    }

    public boolean contains(ResourceEntry rl) {
        return registry.containsKey(rl);
    }

    @SubscribeEvent
    public void onRegistryDump() {
        System.out.println("Registry dump: " + type.getSimpleName());
//        System.out.println(hashCode());
        for (Map.Entry<ResourceEntry, T> entry : entries()) {
            T object = entry.getValue();
            ResourceEntry rl = entry.getKey();

            System.out.println("  (" + rl + ") -> " + object);
        }
    }

    public void register(Class<? extends ObjectInit<T>> clazz, String addonId) {
        // Get fields.
        Field[] fields = clazz.getDeclaredFields();

        System.out.println("Registering object-initialization class: " + clazz);

        // Loop fields.
        for (Field field : fields) {
            // Try.
            try {
                if (type.isAssignableFrom(field.getType())) {
                    // Get register-object.
                    T object = (T) field.get(null);

                    // Set key.
                    if (object.getRegistryName() == null) {
                        if (object.isTempRegistryName()) {
                            object.updateRegistryName(addonId);
                        } else {
                            object.setRegistryName(new ResourceEntry(addonId, field.getName().toLowerCase()));
                        }
                    }

                    // Register if it isn't.
                    if (!values().contains(object)) {
                        register(object.getRegistryName(), object);
                    }
                }
            } catch (IllegalAccessException e) {
                // Oops, some problem occurred.
                e.printStackTrace();
            }
        }
    }

    /**
     * Register an object.
     *
     * @param rl  the resource location.
     * @param val the register item value.
     */
    public void register(ResourceEntry rl, T val) {
        if (!type.isAssignableFrom(val.getClass())) {
            throw new IllegalArgumentException("Not allowed type detected, got " + val.getClass() + " expected assignable to " + type);
        }

        registry.put(rl, val);
    }

    public Collection<T> values() {
        return Collections.unmodifiableCollection(registry.values());
    }

    public Set<ResourceEntry> keys() {
        return Collections.unmodifiableSet(registry.keySet());
    }

    public Set<Map.Entry<ResourceEntry, T>> entries() {
        // I do this because IDE won's accept dynamic values ans keys.
        ArrayList<T> values = new ArrayList<>(values());
        ArrayList<ResourceEntry> keys = new ArrayList<>(keys());

        if (keys.size() != values.size()) throw new IllegalStateException("Keys and values have different lengths.");

        Set<Map.Entry<ResourceEntry, T>> entrySet = new HashSet<>();

        for (int i = 0; i < keys.size(); i++) {
            entrySet.add(new AbstractMap.SimpleEntry<>(keys.get(i), values.get(i)));
        }

        return Collections.unmodifiableSet(entrySet);
    }

    public Class<T> getType() {
        return type;
    }

    @Nullable
    @Deprecated
    public EventBus.Handler getHandler() {
        return null;
    }

    public void registrable(ResourceEntry rl, RegistryEntry object) {
        if (type.isAssignableFrom(object.getClass())) {
            register(rl, (T) object);
        }
    }
}

File: src/main/kotlin/com/qtech/bubbles/registry/Registry.kt
package com.qtech.bubbles.registry

import com.qtech.bubbles.common.IRegistryEntry
import com.qtech.bubbles.common.RegistryEntry
import com.qtech.bubbles.common.ResourceLocation
import com.qtech.bubbles.common.init.ObjectInit
import com.qtech.bubbles.common.maps.SequencedHashMap
import com.qtech.bubbles.event.Bus.qBubblesEventBus
import com.qtech.bubbles.event.SubscribeEvent
import java.util.*
import java.util.AbstractMap.SimpleEntry

open class Registry<T : IRegistryEntry?> protected constructor(clazz: Class<T>, registryName: ResourceLocation) {
    private val registry = SequencedHashMap<ResourceLocation, T>()
    val type: Class<T>
    val registryName: ResourceLocation

    /**
     * Returns the registered instance from the given [ResourceLocation]
     *
     * @param key the namespaced key.
     * @return an registered instance of the type [T].
     * @throws ClassCastException if the type is invalid.
     */
    operator fun get(key: ResourceLocation): T? {
        require(registry.containsKey(key)) { "Cannot find object for: " + key + " | type: " + type.simpleName }
        return registry[key]
    }

    operator fun contains(rl: ResourceLocation?): Boolean {
        return registry.containsKey(rl)
    }

    @SubscribeEvent
    fun onRegistryDump() {
        println("Registry dump: " + type.simpleName)
        //        System.out.println(hashCode());
        for ((rl, `object`) in entries()) {
            println("  ($rl) -> $`object`")
        }
    }

    @Suppress("UNCHECKED_CAST")
    fun register(clazz: Class<out ObjectInit<T>?>, addonId: String?) {
        // Get fields.
        val fields = clazz.declaredFields
        println("Registering object-initialization class: $clazz")

        // Loop fields.
        for (field in fields) {
            // Try.
            try {
                if (type.isAssignableFrom(field.type)) {
                    // Get register-object.
                    val `object` = field[null] as T

                    // Set key.
                    if (`object`!!.registryName == null) {
                        if (`object`.isTempRegistryName) {
                            `object`.updateRegistryName(addonId!!)
                        } else {
                            `object`.registryName = ResourceLocation(addonId, field.name.lowercase())
                        }
                    }

                    // Register if it isn't.
                    if (!values().contains(`object`)) {
                        register(`object`.registryName!!, `object`)
                    }
                }
            } catch (e: IllegalAccessException) {
                // Oops, some problem occurred.
                e.printStackTrace()
            }
        }
    }

    /**
     * Register an object.
     *
     * @param rl  the resource location.
     * @param val the register item value.
     */
    fun register(rl: ResourceLocation, `val`: T) {
        registry[rl] = `val`
    }

    fun values(): MutableCollection<T> {
        return Collections.unmodifiableCollection(registry.values)
    }

    fun keys(): MutableSet<ResourceLocation> {
        return Collections.unmodifiableSet(registry.keys)
    }

    fun entries(): Set<Map.Entry<ResourceLocation, T>> {
        // I do this because IDE won's accept dynamic values ans keys.
        val values = ArrayList(values())
        val keys = ArrayList(keys())
        check(keys.size == values.size) { "Keys and values have different lengths." }
        val entrySet: MutableSet<SimpleEntry<ResourceLocation, T>> = HashSet()
        for (i in keys.indices) {
            entrySet.add(SimpleEntry(keys[i], values[i]))
        }
        return Collections.unmodifiableSet(entrySet)
    }

    @Suppress("UNCHECKED_CAST")
    fun registrable(rl: ResourceLocation, `object`: RegistryEntry) {
        if (type.isAssignableFrom(`object`.javaClass)) {
            register(rl, `object` as T)
        }
    }

    companion object {
        private val metricsMap = HashMap<Class<*>, Registry<*>>()
        fun <T : RegistryEntry?> create(clazz: Class<T>, registryName: ResourceLocation): Registry<T> {
            return Registry(clazz, registryName)
        }

        @Suppress("UNCHECKED_CAST")
        fun <T : RegistryEntry?> getRegistry(objType: Class<T>): Registry<T> {
            return metricsMap[objType] as Registry<T>
        }
    }

    init {
        check(!metricsMap.containsKey(clazz))
        this.registryName = registryName
        type = clazz
        qBubblesEventBus.register(this)
        metricsMap[type] = this
    }
}